README.md
# js-coroutines

[![Logo](http://js-coroutines.com/splash.png)](http://js-coroutines.com)

   [![Stargazers repo roster for js-coroutines](https://reporoster.com/stars/miketalbot/js-coroutines)](https://github.com/miketalbot/js-coroutines/stargazers)


**Supports all browsers and React Native**

When is the right time to sort a massive array on the main thread of a Javascript app? Well any time you
like if you don't mind the user seeing all of your animations and effects jank to hell. Even transferring
to a worker thread is going to hit the main thread for serialization and stutter everything.

So when is the right time? Well it's in all those gaps where you animation isn't doing anything and the
system is idle. If only you could write something to use up that time and then relinquish control to the
system so it can animate and do the rest of the work, then resume in the next gap. Well now you can...

**Get 60fps while sorting an array of 10 million items with _js-coroutines_**

## Quick Start

The project's main web site contains examples of js-coroutines in operation, explains how it can provide
benefits to your project and has links to the full API docs plus some examples.

[JS-COROUTINES Overview and API docs](http://js-coroutines.com)


## How it works?

[This dev.to article goes into detail about how js-coroutines works](https://dev.to/miketalbot/60fps-javascript-while-stringfying-and-parsing-100mbs-of-json-84l)

## Demo

See the [Code Sandbox Demo](https://codesandbox.io/s/js-coroutines-json-demo-etyft?file=/src/App.js).

## Animating Using Coroutines

Another super useful way of using coroutines is to animate and control complex states - js-coroutines provides this too with the powerful 
`update` method that runs every frame in high priority. 

There's an example of how to write your own animation later and you 
can see [this CodeSandbox demo](https://codesandbox.io/s/coroutines-examples-zeq33) of stateful animations, or [this game built using js-coroutines](https://codesandbox.io/s/condescending-waterfall-v8mxq), for more.




## Commonly required asynchronous operations

You can use `*stringify()` and `*parse()` to manipulate JSON in an idle coroutine that won't
block the main thread.

You can use `stringifyAsync()` and `parseAsync()` to perform JSON parsing and stringifying
anywhere you can take a promise or `await` a response.

You can use `*compress()` and `*decompress()` to compress to storable/transmittable strings.

You can use `compressAsync()` and `decompressAsync()` to perform compression and decompression
anywhere you can take a promise or `await` a response.


### Compression

js-coroutines uses lz-string for compression.

[LZ-String GitHub/Documentation](https://github.com/pieroxy/lz-string).

## Installation

```sh
npm install --save js-coroutines
```

## Usage

You can make your own generator functions that do anything you like and `yield` to check
if there is time remaining this frame:

```js
import {run, sort, stringify} from 'js-coroutines'

...

let json = await run(function*() {
  const results = [];
  for(let i = 0; i < 10000000; i++) {
    results.push(Math.random() * 10000);
    //Check how much time left every 100 entries
    if(i % 100 === 0) yield;
  }
  //Pass to a coroutine sort function
  yield* sort(results, value=>value)
  return yield* stringify(results);
})

```

Or you can just use the Async helper functions in an async routine. This is
less powerful, but you don't have to start writing generator functions
or working out where to yield.

```js
import { parseAsync, mapAsync } from "js-coroutines";

async function process(url) {
  const response = await fetch("someurl");
  //Use the coroutine version of parse, rather than blocking
  //the main thread permanently by using .json()
  const result = await parseAsync(await response.text());
  //Imagining the result is some database rows, map out the
  //desired response without blocking the main thread for paints
  const values = await mapAsync(result, (row) => ({
    item: row.time,
    value: row.quantity * row.unitPrice,
  }));
  return values;
}
```


## Getting Started With Async Functions

Async functions are the easiest way to use js-coroutines if you just need to
handle common functions like sorts, finds, filters and JSON parsing in the
background. If you need to break up your own logic you will have to write
a generator.

Just import the `xxxAsync` version of the function from js-coroutines and
use a standard Promise chain or `await` and the code will run only in the
gaps.

```js
async function asyncFunctions() {
  // Parse the JSON async
  let o = await parseAsync(json);
  // Concatenate arrays in the background
  for (let i = 1; i < 12; i++) {
    o = await concatAsync(o, o);
  }
  // Write out the arrays
  let output = await stringifyAsync(o);
  // Map ids from the array in the background
  let justIds = await mapAsync(o, (v) => v.id);
  // Return the JSON of just the ids
  return [output, await stringifyAsync(justIds)];
}
```

## Getting Started With Function Pipelines

You can also create pure functional pipelines using pipe, tap, branch, repeat and call

```js
      import {pipe, parseAsync, tap, mapAsync} from 'js-coroutines'

      const process = pipe(
             parseAsync,
             function * (data) {
                let i = 0
                let output = []
                for(let item of data) {
                    output.push({...item, total: item.units * item.price})
                    if((i++ % 100)==0) yield
                }
                return output
             },
             mapAsync.with(v=>({value: v.total, item: v.item})),
             tap(console.log),
             stringifyAsync
         )
         
      ...
      
      console.log(await process(data))
 ```               
                    


## Getting Started Writing Your Own Generators

`js-coroutines` uses generator functions and `requestIdleCallback` to let you easily split up
your work with minimal effort.

A simple generator:

```js
await run(function* () {
  const strings = [];
  let results;

  //Create 2 million rows of random values
  results = new Array(2000000);
  for (let i = 0; i < 2000000; i++) {
    //Every 128th record, check to see if we still have time
    //run the remainder on another tick if we don't
    if ((i & 127) === 0) yield;
    results[i] = (Math.random() * 10000) | 0;
  }

  //Double all the values
  yield* forEach(
    results,
    yielding((r, i) => (results[i] = r * 2))
  );

  //Get the square roots
  const sqrRoot = yield* map(
    results,
    yielding((r) => Math.sqrt(r))
  );

  //Sum all of the items
  const sum = yield* reduce(
    results,
    yielding((c, a) => c + a, 64),
    0
  );

  //Join the arrays
  yield* append(results, sqrRoot);

  // Sort the results
  yield* sort(results, (a, b) => a - b);
  return results;
});
```

As you can probably see, it comes ready with the most useful functions for arrays:

- `forEach`
- `map`
- `filter`
- `reduce`
- `findIndex`
- `find`
- `some`
- `every`
- `sort`
- `append` (array into array)
- `concat` (two arrays into a new array)

The helper `yielding` wraps a normal function as a generator and checks remaining time
every few iterations. You can see it in use above. It's just a helper though - if
your `map` function needs to do more work it can just be a generator itself,
yield when it likes and also pass on to deeper functions that can yield:

```js
const results =
  yield *
  map(inputArray, function* (element, index) {
    //Every 200 indices give up work
    //on this frame by yielding 'true'
    //yield without true, checks the amount
    //of remaining time
    if (index % 200 === 199) yield true;

    //Yield out a filter operation
    let matched = yield* filter(
      element,
      yielding((c) => c > 1000)
    );

    //Now yield out the calculation of a sum
    return yield* reduce(
      matched,
      yielding((c, a) => c + a),
      0
    );
  });
```

`yielding(fn, [optional yieldFrequency]) -> function *`

##

## Async

Former `runAsync` is deprecated. You may yield a Promise instead.
js-coroutines will automatically restart the coroutine when the
Promise is resolved.

```js
const results = await run( function* () {
  const response = yield fetch("http://someurl");
  const rows = yield response.json();
  yield* sort(rows, (a) => a.value);
  return processed;
});
```

# Update coroutines

A great way to do stateful animation is using a coroutine running every frame.
In this case when you `yield` you get called back on the next frame making
stateful animations a piece of cake:

```js
import { update } from "js-coroutines";

//Animate using a coroutine for state
update(function* () {
  while (true) {
    //Move left to right
    for (let x = -200; x < 200; x++) {
      logoRef.current.style.marginLeft = `${x * multiplier}px`;
      yield;
      //Now we are on the next frame
    }
    //Move top to bottom
    for (let y = 0; y < 200; y++) {
      logoRef.current.style.marginTop = `${y * multiplier}px`;
      yield;
    }
    //Move diagonally back
    for (let x = 200; x > -200; x--) {
      logoRef.current.style.marginLeft = `${x * multiplier}px`;
      logoRef.current.style.marginTop = ((x + 200) * multiplier) / 2 + "px";
      yield;
    }
  }
});
```

# Writing Coroutines with the API

### `run(coroutineFunction, msToLeaveSpare=1, timeout=160) -> TerminatablePromise(Any)`

`coroutineFunction` must be a `function *`

Run your coroutine, which will occupy up to the last amount of
m/s specified in the `msToLeaveSpare` (0.5 is the minimum) of the idle
time on the thread. `timeout` specifies the time before it will run
if there is no idle time (default 1/10 frames).

The promise returned has a `terminate(result)` function that can be used
to stop the calculation early - maybe you want to go again with different
parameters.

`yield` inside your coroutine will check how much time is left and continue
if there is enough.

`yield 2` yielding a number results in a check for at least that number of ms remaining.
 
`yield true` will definitely abandon the current frames work. Useful if you
are about to/just have allocated tons of memory to give time for GC.

`yield* generatorFn([param], [...param])` call a generator function which
will take over yielding time checks and return the value it creates when done.

```js
function* myCoroutine() {
  const results = [];
  for (let i = 1; i < 1000000; i++) {
    if ((i & 127) === 0) yield; //time check
    results.push(i);
  }
  yield true; // end current frame processing
  let anotherArray = new Array(results.length);
  yield true; // give time for GC
  // Run a for loop on the results
  yield* forEach(
    results,
    yielding(
      (result, index, collection) =>
        (anotherArray[index] = result / collection.length)
    )
  );
  return anotherArray;
}
```

### `*yielding(fn, [optional frequency=8]) -> function *`

Converts a normal function into one that yields every `frequency` calls.

Very useful for providing map/filter functions etc.

### `wrapAsPromise(coroutine) -> function([params]) -> Promise(Any)`

Returns an async function that can be called with await and will call
the passed in coroutine forwarding parameters.

```js
//Create an async function
const toTuplesAsync = wrapAsPromise(function* (array) {
  let output = [];
  //Create tuples
  for (let i = 0; i < array.length; i += 2) {
    output.push([array[i], array[i + 1]]);
    yield;
  }
  return output;
});

...

async function myProcess() {
  const data = await getDataFromSomewhere();
  //Call your wrapped coroutine
  const tuples = await toTuplesAsync(data);
  //do something with the result
  return processTuplesSomehow(tuples);
}


```

## License

js-coroutines - MIT (c) 2020 Mike Talbot

Timsort - MIT (c) 2015 Marco Ziccardi (c) 2020 Mike Talbot (Generator modifications)

JSON stringify - Public Domain (c) 2017 Douglas Crockford (c) 2020 Mike Talbot (Generator modifications)

JSON Parse - yastjson - MIT (c) 2020 5u9ar (zhuyingda) (c) 2020 Mike Talbot (Optimisations and generator modifications)


CHANGELOG.md
### 2.4.36

-   Split out core coroutine functions to reduce bundle sizes

### 2.4.35

-   Make @babel/runtime a peer dep

#### 2.4.32

-   Make all functions individually importable

#### 2.4.30

-   Update NPM license info

#### 2.4.28

-   NodeJS idles more when not busy


#### 2.4.26

-   Enable yielding functions to return a promise

#### 2.4.25

-   Split out async wrapped functions and add webpack sideEffects

#### 2.4.24

-   Tune timeout parameters
-   Animation balancing

#### 2.4.22

-   Change looping order for conditions where multiple idle callbacks overlap
-   Remove default timeout for idles as it tends to cause animation glitches

#### 2.4.21

-   Adjust timing

#### 2.4.19

-   Provide simpler default sort comparison

#### 2.4.18

-   Reduce timeout overlaps

#### 2.4.17

-   Optimise sort algorithm

#### 2.4.15

-   Tune default sequence yields

#### 2.4.13

-   Fix minification bugs

#### 2.4.12

-   Additional package size reduction

#### 2.4.11

-   Reduce package size of build

#### 2.4.10

-   Merge PR1/2

#### 2.4.9

-   Fix NodeJS version 

#### 2.4.8

-   Fix issue with stringifying empty strings (#10)
-   Polyfill performance.now() (#9)
-   Fix issue with parsing JSON including whitespace (#7)

#### 2.4.7

-   Update dependencies

#### 2.4.6

-   Update dependencies for audit fix

#### 2.4.5

-   Use setImmediate for NodeJS as it has better performance

#### 2.4.4

-   Another change, for remote trials

#### 2.4.3

-   Tweak the amount of delay NodeJS allows

#### 2.4.2

-   Fix some Node issues
-   Fix a test issue

#### 2.4.1

-   Added NodeJS support
-   sortAsync/sort return the array for chaining
-   Uint8Array fixes for JSON
-   Removed support for array replacer in stringify
-   Added tests using new NodeJS runner
-   several bug fixes

#### 2.3.72

-   Joined promises are chainable

#### 2.3.71

-   Make additional promises only cancel on termination

#### 2.3.70

-   Fix typings
-   Fix reduceAsync

#### 2.3.69

-   Add a join for multi-chaining terminated promises from a singleton
-   Improve typings for singleton and terminatable promises

#### 2.3.68

-   Add a singleton() function wrapper for an easy way to have a cancellable version of a coroutine.
    Coroutines wrapped in a singleton are run as an async function returning a Promise,
    but if the previous one hasn't completed it immediately returns with a default value.

-   Improve documentation and add useInternalEngine to typings

#### 2.3.66

-   Add useInternalEngine(true) call to enable the polyfilled version in all circumstances

#### 2.3.65

-   Enable the override of standard yielding function for asyncs

#### 2.3.64

-   Change loop time calculator back to api

#### 2.3.63

-   Remove the `debugger` - oops

#### 2.3.62

-   Update docs

#### 2.3.61

-   Added functionality for keyBy, uniqueBy, indexOf, lastIndexOf, groupBy, includes
-   Made filter/every/some/forEach handle objects and arrays (per lodash)

#### 2.2.60

-   enabled falsey values in the pipeline to make composition easier

#### 2.2.59

-   Improved performance tracking in idle
-   Added pipe/tap/branch/call/repeat to the API for functional pipelines
-   Fix some JSON issues (note JSON parse seen NOT working with optional whitespace still)

#### 2.1.56

-   Fix a bug with reduce/reduceAsync
-   try to get typings working again

#### 2.1.54

-   Move to transform runtime to remove regenerator issues

#### 2.1.53

-   Ensure requestAnimationFrame starts on next frame

#### 2.1.52

-   Buffer animation callbacks to reduce calls to requestAnimationFrame

#### 2.1.51

-   Fix a bug in starting update coroutines with parameters

#### 2.1.50

-   Handle the case where the Promise yielded is immediately resolved without
    going through the restart process

#### 2.1.49

-   Added the ability to yield a Promise and have the coroutine
    restart when it finishes

-   Deprecated runAsync (due to above)

#### 2.1.44 - 2.1.48

-   Adjust polyfill for a variety of Safari use cases

#### 2.1.43

-   Fix polyfill on Safari

#### 2.1.42

-   Can start `run` or `update` with the results of calling a generator to enable parameter passing
-   Main website linked

#### 2.1.41

-   Improved polyfill

#### 2.1.40

-   Tested react native

#### 2.1.38

-   Provide own polyfill for requestIdleCallback

#### 2.1.37

-   remove direct reference to requestIdleCallback polyfill for React Native support. This
    can now be added manually on websites requiring it
-   first attempts at RN support

#### 1.1.36

-   fix documentation typo

#### 1.1.35

-   fixed build issue

#### 1.1.34

-   Update documentation
-   Fix an issue with encodeURIComponent version of lz-string

#### 1.1.33

-   Add compress and decompression from lz-string in the gaps

#### 1.1.32

-   setTimeout idle time background processing
-   can yield a number to request that number of m/s in the current frame (request does not carry over to the next frame)
-   added async examples to getting started

#### 1.1.31

-   Too early for last update, reverting

#### 1.1.30

-   Use setTimeout for run timeouts to ensure we get some background processing

#### 1.1.29

-   Fix a bug where terminate on the run promise was lost due to double async wrapping

#### 1.1.28

-   Massively improve performance by replacing Yastjson's stack based bracket analyser with a couple of ints

#### 1.1.27

-   Improve performance by reducing yields during the tokenise phase to 1 in every 2048 characters

#### 1.1.26

-   Rewrite JSON tokenizer and AST for performance (again!)

#### 1.1.25

-   Fix a problem with escaped JSON strings not being fully handled by yastjson
-   Reintroduce .toJSON() in stringify

#### 1.1.24

-   Made sure Async functions return a terminatable promise
-   Fixed the CHANGELOG formatting!

#### 1.1.23

-   Fix a bug with resolver now just being the actual promise resolve.

#### 1.1.22

-   Link article on how it works

#### 1.1.21

-   Documentation improvements

#### 1.1.20


src/App.js
import React, {useEffect, useRef} from 'react'
import logo from './logo.svg'
import './App.css'
import {
    append,
    compressAsync,
    concatAsync,
    decompressAsync,
    every,
    forEach,
    groupByAsync,
    keyByAsync,
    map,
    mapAsync,
    parse,
    parseAsync,
    pipe,
    reduce,
    repeat,
    singleton,
    sort,
    stringify,
    stringifyAsync,
    tap,
    update,
    yielding
} from './component'
import isEqual from 'lodash/isEqual'
import test from './component/yastjson/test/test.json'

const {format} = new Intl.NumberFormat()

const json = JSON.stringify(test)
const referenceObject = JSON.parse(json)

const job = singleton(function* (add) {
    let results
    results = new Array(2000000)
    for (let i = 0; i < 2000000; i++) {
        if ((i & 127) === 0) yield
        results[i] = (Math.random() * 10000) | 0
    }
    add(`Created ${format(results.length)} items`)
    //Double all the values
    yield* forEach(
        results,
        yielding((r, i) => (results[i] = r * 2))
    )
    add(`Doubled the value of ${format(results.length)} items`)

    const result = yield new Promise((resolve) => {
        setTimeout(() => {
            resolve(1234)
        }, 1000)
    })
    add(`Result of a Promise was ${result}`)
    //Get the square roots
    const sqrRoot = yield* map(
        results,
        yielding((r) => Math.sqrt(r))
    )
    add(`Created a new array with the square roots of ${format(sqrRoot.length)} items`)
    const asAString = yield* stringify(results)
    add(`Stringified ${format(sqrRoot.length)} items to JSON string length ${format(asAString.length)}`)
    const copy = yield* parse(asAString)
    add(`Parsed JSON back to a copy of ${format(copy.length)} array items`)

    const ok = yield* every(
        copy,
        yielding((v, i) => results[i] === v)
    )
    add(`Validated that ${format(copy.length)} parsed values matched - "${ok ? 'ok' : 'mismatch'}"`)

    for (let i = 0; i < 1000; i++) {
        yield* parse(json)
    }
    const obj = yield* parse(json)
    if (!isEqual(obj, referenceObject)) {
        console.log(obj, referenceObject)
        add(`Parsed example JSON back to an object 1000 times - "error"`)
    } else {
        add(`Parsed example JSON back to an object 1000 times - "matched"`)
    }

    add(
        `Sum of ${format(results.length)} items is ${format(
            yield* reduce(
                results,
                yielding((c, a) => c + a),
                0
            )
        )}`
    )
    //Join the arrays
    yield* append(results, sqrRoot)
    add(`Appended the square roots to the normal values making ${format(results.length)} items in the array`)
    // Sort the results
    yield* sort(results, (a, b) => a - b)
    add(`Sorted ${format(results.length)} items`)
    return results
}, 'test')

function App() {
    const strings = []
    const [process] = React.useState(() =>
        pipe(
            parseAsync,
            tap((o) => add(`PIPE: parsed JSON to create an array of ${format(o.length)} items`)),
            repeat((i) => concatAsync(i, i), 11),
            tap((o) => add(`PIPE: doubled array 12 times to an array of ${format(o.length)} items`)),
            stringifyAsync,
            tap((o) => add(`PIPE: stringified to a string ${format(o.length)} chars`)),
            compressAsync,
            tap((o) => add(`PIPE: compressed JSON to a string of ${format(o.length)}`)),
            decompressAsync,
            tap((o) => add(`PIPE: decompressed data to a string of ${format(o.length)}`)),
            parseAsync,
            tap((o) => add(`PIPE: hydrated ${format(o.length)} items`)),
            mapAsync.with((v) => v._id),
            tap((o) => add(`PIPE: mapped the ids from objects in the array to a new array of ${format(o.length)}`)),
            stringifyAsync,
            tap((o) => add(`PIPE: new array to a string of length ${format(o.length)}`))
        )
    )

    const logoRef = useRef()
    const outputRef = useRef()
    useEffect(() => {
        // test();
        animate().then(() => console.log('Done'))
        calculateAsync().then(() => console.log('Calculation Done'))
        asyncFunctions().then(() => console.log('Async done'))
        process(json).then(()=>console.log("Pipe done"))
    })
    return (
        <div className="App">
            <header className="App-header">
                <div ref={logoRef} style={{ position: 'relative', marginLeft: 0 }}>
                    <img src={logo} className="App-logo" alt="logo" />
                </div>
                <div className="info-list">
                    <ul ref={outputRef}></ul>
                </div>
            </header>
        </div>
    )

    function animate() {
        let multiplier = window.innerWidth / 300
        return update(function* () {
            while (true) {
                for (let x = -200; x < 200; x++) {
                    logoRef.current.style.marginLeft = `${x * multiplier}px`
                    yield
                }
                for (let y = 0; y < 200; y++) {
                    logoRef.current.style.marginTop = `${y * multiplier}px`
                    yield
                }
                for (let x = 200; x > -200; x--) {
                    logoRef.current.style.marginLeft = `${x * multiplier}px`
                    logoRef.current.style.marginTop = `${((x + 200) * multiplier) / 2}px`
                    yield
                }
            }
        })
    }

    async function asyncFunctions() {
        //Parse the JSON async
        let o = await parseAsync(json)
        add(`ASYNC: parsed JSON to array of ${format(o.length)} items`)
        for (let i = 1; i < 12; i++) {
            o = await concatAsync(o, o)
        }
        add(`ASYNC: doubled the array 12 times to an array of ${format(o.length)} items`)
        let output = await stringifyAsync(o)
        add(`ASYNC: stringified array to a ${format(output.length)} character string`)
        let compressed = await compressAsync(output)
        add(`ASYNC: compressed the output to a string of ${format(compressed.length)} characters`)
        let decompressed = await decompressAsync(compressed)
        add(
            `ASYNC: decompressed the output to a string of ${format(decompressed.length)} characters - ${
                decompressed === output ? 'Matches' : "Didn't match"
            }`
        )
        let justIds = await groupByAsync(o, (v) => {
            return v._id
        })
        let index = await keyByAsync(o, v=>v._id)
        console.log({index})
        add(`ASYNC: mapped the ids from objects in the array to an index of ${format(Object.keys(justIds).length)} items`)
        output = await stringifyAsync(justIds)
        add(`ASYNC: stringified the index to a ${format(output.length)} character string`)


    }

    async function calculateAsync() {
        job(add).then(v=>add(v==='test' ? "TEST SINGLETON PASSED" : 'TEST SINGLETON FAILED'))
        return await job(add)
    }

    function add(message) {
      console.log(message)
        requestAnimationFrame(() => {
            strings.push(`<li>${message}</li>`)
            if (outputRef.current) {
                outputRef.current.innerHTML = strings.join('')
            }
        })
    }
}

export default App
